Lib/http/server.py

```diff
@@ -892,21 +933,34 @@ def translate_path(self, path):
             path += '/'
         return path
 
-    def copyfile(self, source, outputfile):
-        """Copy all data between two file objects.
+    def copyfile(self, source, outputfile, *, range=None):
+        """Copy all data between two file objects if range is None.
+        Otherwise, copy data between two file objects based on the
+        inclusive range (start, end).
 
         The SOURCE argument is a file object open for reading
-        (or anything with a read() method) and the DESTINATION
-        argument is a file object open for writing (or
-        anything with a write() method).
+        (or anything with read() and seek() method) and the
+        DESTINATION argument is a file object open for writing
+        (or anything with a write() method).
 
         The only reason for overriding this would be to change
         the block size or perhaps to replace newlines by CRLF
         -- note however that this the default server uses this
         to copy binary data as well.
 
         """
-        shutil.copyfileobj(source, outputfile)
+        if range is None:
+            shutil.copyfileobj(source, outputfile)
+        else:
+            start, end = range
+            length = end - start + 1
+            source.seek(start)
+            while length > 0:
+                buf = source.read(min(length, shutil.COPY_BUFSIZE))
+                if not buf:
+                    raise EOFError('File shrank after size was checked')
+                length -= len(buf)
+                outputfile.write(buf)
```

Comment:
Once we have a multi-range, this will be called for each of the range we constructed. So I think this part will need to be a private method:

```
def _copyfile(self, out, src, size):
    ...
```

and it will be used as

```
for start, end in ranges:
    src.seek(start)
    self._copyfile(self, out, src, end - start + 1)
```

However, for multi-ranges, we'll likely need to rename `range` into `ranges` and accept an iterable of "ranges". As such, it might be better to already accept such iterables and reject those that are not of length 1 for now.

---

**Alternatively**, and this could be perhaps better, we can implement `copyfile` only for a single range, and be responsible to call it for all range items for multi-range.